tools/purge_ems.rb

```diff
@@ -0,0 +1,109 @@
+#!/usr/bin/env ruby
+require File.expand_path('../config/environment', __dir__)
+require 'optimist'
+
+ARGV.shift if ARGV[0] == '--' # if invoked with rails runner
+batch_message = "For each relationship, the number of rows to be removed in a batch. A lower number may decrease memory usage but also take longer."
+opts = Optimist.options do
+  banner "Purge management system records in the background via the queue.\n\nUsage: ruby #{$PROGRAM_NAME} [options]\n\nOptions:\n\t"
+  opt :id,      "Management System id",                                         :type => :integer,                    :required => true
+  opt :follow,  "Follow the progress or exit after creating work items.",       :type => :boolean, :default => true
+  opt :batch,   batch_message,                                                  :type => :integer, :default => 1_000
+  opt :timeout, "The number of minutes each work item will be allowed to run.", :type => :integer, :default => 30
+end
+
+Optimist.die :id,      "must be a positive number"                             if opts[:id] < 1
+Optimist.die :batch,   "must be a positive number"                             if opts[:batch] < 1
+Optimist.die :timeout, "must be a positive number greater than or equal to 10" if opts[:timeout] < 10
+
+DESTROY_METHOD = "destroy".freeze
+ZONE           = nil                    # Will queue for any zone
+PRIORITY       = MiqQueue::LOW_PRIORITY # Don't block other work
+
+def log(msg)
+  $log.info("MIQ(#{__FILE__}) #{msg}")
+  puts msg
+end
+
+def current_backlog
+  MiqQueue.where(:method_name => DESTROY_METHOD, :zone => ZONE, :priority => PRIORITY).count
+end
+
+id = opts[:id]
+ems = ExtManagementSystem.find(id)
+if ems.enabled?
+  log("Pausing management system with id: #{id} #{ems.name}...")
+  ems.pause!
+  log("Pausing management system with id: #{ems.id} #{ems.name}...Complete")
+end
+
+log("Removing any workers for management system with id: #{id} #{ems.name}...")
+ems.ems_workers.each(&:kill_async)
```

Comment:
WIll this require that the script is run from inside the orchestrator?  Otherwise how will the code communicate with kubernetes to stop the workers?

Reply:
@jrafanie Can we just remove this entire pausing logic from the script and just check up front if the provider is disabled?  The instructions for running the script can then just be "ensure you have paused the EMS in question" and the user can go to the UI and pause it?  (In actuality if it's a failed deletion attempt, the EMS is likely _already_ paused since that's not the part that gets stuck, so here it can just be a "double check else exit" thing.

Reply:
> WIll this require that the script is run from inside the orchestrator? Otherwise how will the code communicate with kubernetes to stop the workers?

doesn't it queue a message for the server/orchestrator?

Reply:
> doesn't it queue a message for the server/orchestrator?

Yes

Reply:
> Can we just remove this entire pausing logic from the script and just check up front if the provider is disabled?

Why?  The intention was to bring orchestrate_destroy to a script that knows how to batch the relationship destroy before actually trying the ems destroy.

Reply:
And of course, I'm hoping we can move some of the logic into the orchestrate_destroy so it can properly slice up the destroy into smaller batches so it's doable without bloating memory or exceeding timeouts.

Reply:
> The instructions for running the script can then just be "ensure you have paused the EMS in question" and the user can go to the UI and pause it?

The user cannot go into the UI and pause it

Reply:
I forgot about the UI not supporting pause yet.  :-(

I guess we have to pause at least for now.